### DISCUSSION TITLE
Flash memory

### DISCUSSION THREAD

**Discussion by mauroForlimpopoli** [Moddable-OpenSource/moddable#620] [Q&A]
This is my code:
let file = new File(config.file.root + "test.txt", true);
file.position = 0;
file.write("0123456789");
file.close();

in next instruction I will change the file.position = 10 and so on....

The flash memory has a number of writing of about 10.000 times. (I don't understanf if for Block Size, Page Size or single location).
This is my question:  
How many cycles of writing (of 10 byte) can I do if the my SPIFFS file system if about 100kB ? 
every string is 10Byte
so before to complete the memory = 100.000 : 10 = 10.000
10.000 * 10.000(cycles of life) = 10.0000.000 .
Is it correct ?

**Comment by phoddie** [comment]
A flash page is 4096 bytes. The limit of flash isn't the number of writes, but the number of erases. Your 10 byte string will fit 409 times into a single page after a single erase. If the limit is 10,000 erases, you can reliably write about 4,000,000 values to a single page. You have a 100 KB file, which is 25 pages, so  it would be 25 * 4,000,000 = 100,000,000 values written.

The number of erases is likely higher. It depends on the chip. Also, SPIFFs can be configured for wear-leveling, which avoids reusing the same page so that the erased are spread out, when there is free space.

**Reply by mauroForlimpopoli** [reply]
Something is not completly clear for me. If I write until the file position is 100 and then I set the position to 10, it writes my new 10 byes over that I wrote before. So why I need to erase the block written yet? How can I delete a single page of my file ?
regards

**Reply by phoddie** [reply]
You are using a file system API. It takes care of erasing when needed. You don't need to call erase.

**Reply by mauroForlimpopoli** [reply]
Perfect. Thank you

**Reply by mauroForlimpopoli** [reply]
Now I created a file of 500kB . How can I read it? The RAM available is not enought , is there a file.position for the end too?
I  tried this:
let buffer7;	
file.position = 0;
	while (file.position < 1020) {
		buffer7 = file.read(ArrayBuffer, 1024);
	}
trace(buffer7[0] + "\n") 
the output is always undefined
end then:
let buffer7;	
file.position = 0
	while (file.position < 1024) {
		buffer7 = file.read(String);
	}
trace(buffer7 +"\n");
the output is memory full.
in both the case the system byte free is about 60kB just before of "while (file.position < 1024) {";

**Comment by phoddie** [comment]
You wrote the file in parts. You can read the file in parts too. The [documentation](https://github.com/Moddable-OpenSource/moddable/blob/public/documentation/files/files.md#file) for the Files module explains the relevant APIs.

Our [book](https://www.moddable.com/book) has sections on "Reading Text in Pieces" (Page 229) and "Reading Binary Data in Pieces" (page 231). It also has "Getting File Size" (page 228).

Your code to read the file as a String fails because you pass only a single argument, which means "read the entire file." Instead of `file.read(String)` use `file.read(String, 10)`.

This code following your `ArrayBuffer` loop will always print `undefined`:

```js
trace(buffer7[0] + "\n")
```

That is because `buffer7` is an `ArrayBuffer`.  The following also displays `undefined` if you enter it into the console of any web browser:

```js
x = new ArrayBuffer(10);
console.log(x[0]);
```

You cannot read the content of an `ArrayBuffer` directly .You must wrap in a view (TypedArray or DataView) to read and write. That is explained on page 107, beginning with the paragraph "You might expect to be able to access the content of an `ArrayBuffer` using array syntax..".  Here's how you might adapt your line above to print the value of the first byte.

```js
let view7 = new Uint8Array(buffer7));
trace(view7[0] + "\n");
```

**Reply by mauroForlimpopoli** [reply]
Thank you again.
